Dynatrace/dynatrace-operator
# Fix GORM and remove usage of digest

## Description

Initial change:
We no longer need to use the digest to pull the image in case of `CodeModulesImage`, so now we just use the ref we get from the user, and use that string (base64 encoded) as the folder name.

**To make the change actually work:**
I had to update the way we do things all around the new GORM implementation, so that we are utilising the data in the new database schema in a way that "make more sense".
- This includes changes to the GC, now it actually runs when there is more likely some stuff to collect (example: when a dynakube is deleted)
- It actually cleans things up (⚠️ this is very important)
  - The `OSMounts` have to be cleaned up, otherwise the `TenantConfigs` can't be cleaned up
     - So we don't fill the database with bunch of deleted `TenantConfigs` (this is more of a cosmetic problem) 
  - The `CodeModule` foreign-key in an `AppMount` has to be set to Null in case the `AppMount` is deleted, so the `CodeModule` can be deleted if its not used. (Deleted `AppMounts` no longer use the related `CodeModule`) 
    - If a soft deleted `CodeModule` is not cleaned-up soon ™️  then you will have a problem with the `UNIQUE` constraint if the same version would be downloaded.

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ REALLY IMPORTANT NOTES ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ 
- If you have previously installed the Operator that was already using the GORM implementation, you WILL face errors
   - The reason for this is that the current way we use the database is very messy, and things don't get cleaned up (just soft deleted)
   - If the soft-deleted things are not removed, there may be problems due to the `UNIQUE` constraint. (mainly for the `CodeModules`)
   - OSMounts were not updated properly, so they will point to a `VolumeMeta` that is actually not correct anymore

⚠️ ⚠️ ⚠️ ⚠️ ⚠️


## How can this be tested?


Play around with the csi-driver, check that the `CodeModulesImage` works, check that there are no database errors in the logs. 
Check if the garbage-collection works, make sure its not too aggressive.


Sidenote: I still want to add more unit-tests, especially to the "Cleaner" parts, but its ready to be reviewed
